tests/Operation/WatchFunctionalTest.php

```diff
@@ -1190,6 +1197,266 @@ public function testSessionFreed()
+                $changeStream = $operation->execute($this->getPrimaryServer());
+
+                // The first next will hit the fail point, causing a resume
+                $changeStream->next();
+                $changeStream->next();
+            },
+            function(array $event) use (&$aggregateCommands) {
+                $command = $event['started']->getCommand();
+                if ($event['started']->getCommandName() !== 'aggregate') {
+                    return;
+                }
+
+                $aggregateCommands[] = (array) $command;
+            }
+        );
+
+        $this->assertCount(2, $aggregateCommands);
+
+        $aggregateCommands = array_map(
+            function (array $aggregateCommand) {
+                // Remove resume options from the changestream document
+                if (isset($aggregateCommand['pipeline'][0]->{'$changeStream'})) {
+                    $aggregateCommand['pipeline'][0]->{'$changeStream'} = array_diff_key(
+                        (array) $aggregateCommand['pipeline'][0]->{'$changeStream'},
+                        ['resumeAfter' => false, 'startAfter' => false, 'startAtOperationTime' => false]
+                    );
+                }
+
+                // Remove options we don't want to compare between commands
+                return array_diff_key($aggregateCommand, ['lsid' => false, '$clusterTime' => false]);
+            },
+            $aggregateCommands
+        );
+
+        // Ensure options
+        $this->assertEquals($aggregateCommands[0], $aggregateCommands[1]);
+    }
+
+    /**
+     * Prose test: "ChangeStream will not attempt to resume on any error
+     * encountered while executing an aggregate command."
+     */
+    public function testErrorDuringAggregateCommandDoesNotCauseResume()
+    {
+        if (version_compare($this->getServerVersion(), '4.0.0', '<')) {
+            $this->markTestSkipped('failCommand is not supported');
+        }
+
+        $operation = new Watch($this->manager, $this->getDatabaseName(), $this->getCollectionName(), [], $this->defaultOptions);
+
+        $commandCount = [];
+
+        $this->configureFailPoint([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => ['failCommands' => ['aggregate'], 'errorCode' => self::NOT_MASTER],
+        ]);
+
+        $this->expectException(CommandException::class);
+
+        (new CommandObserver)->observe(
+            function() use ($operation) {
+                $operation->execute($this->getPrimaryServer());
+            },
+            function(array $event) use (&$commandCount) {
+                $commandCount++;
+            }
+        );
+
+        $this->assertSame(1, $commandCount);
+    }
+
+    /**
+     * Prose test: "ChangeStream will perform server selection before attempting
+     * to resume, using initial readPreference"
+     */
+    public function testOriginalReadPreferenceIsPreservedOnResume()
```

Comment:
Since this test requires a secondary, you may want to lead with:

```php
try {
    $this->manager->selectServer(new ReadPreference('secondary'));
} catch (ConnectionTimeoutException $e) {
    $this->markTestSkipped('Secondary is not available');
}
```

We do this in DocumentationExamplesTest once for something that uses a secondary read preference. I believe that's the only place, so up to you if you'd like to make a common skip method in FunctionalTestCase for this.

Other instances of secondary-related read preferences are in ClientTest.php, CollectionFunctionalTest, and DatabaseFunctionalTest, but they don't actually require a secondary to exist.

Reply:
I added a `catch` block to the `selectServer` statement below where we select the secondary server for the operation. While we could add a `getSecondaryServer()` method similar to `getPrimaryServer`, I'd rather make skipping the test if no secondary was found explicit instead of hiding it in a helper method.

Reply:
> I'd rather make skipping the test if no secondary was found explicit instead of hiding it in a helper method.

Ah, I was suggesting a `skipIfNoSecondary()` helper, which would be similar to `skipIfChangeStreamIsNotSupported()` or `skipIfTransactionsAreNotSupported()`. I wasn't suggesting a new `getSecondaryServer()` method that skips the test if it fails to select a secondary.

In any event, this change looks good.